<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="MR-Buggy3">
    <material name="red">
        <color rgba="1 0.1 0.1 1"/>
    </material>
    <material name="black">
        <color rgba="0.1 0.1 0.1 1"/>
    </material>
    <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
    </material>
    <material name="yellow">
        <color rgba="1 0.435 0.002 1"/>
    </material>
    <material name="green">
        <color rgba="0.1 1 0.1 1"/>
    </material>
    <!-- 底座 -->
    <link name="footbase">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="100.0" />
            <inertia ixx="20" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="32" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/footbase.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/footbase.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 外壳 -->
    <link name="out_face">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="100.0" />
            <inertia ixx="20" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="12" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/outface.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="yellow" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/outface.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 叉齿架 -->
    <link name="fork_frame">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="200.0" />
            <inertia ixx="40" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="12" />
        </inertial>
        <visual>
            <origin xyz="0 0.0 0.00" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/frame.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/frame.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 叉齿 -->
    <link name="fork">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="100.0" />
            <inertia ixx="50" ixy="0.0" ixz="0.0" iyy="22" iyz="0.0" izz="42" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/fork.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="gray" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/fork.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>

    <!-- 仿撞条 -->
    <link name="bar">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="20.0" />
            <inertia ixx="7" ixy="0.0" ixz="0.0" iyy="3" iyz="0.0" izz="5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/bar.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/bar.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 避障雷达 -->
    <link name="avoid_lidar">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="20.0" />
            <inertia ixx="7" ixy="0.0" ixz="0.0" iyy="3" iyz="0.0" izz="5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/lidar.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="green" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/lidar.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>

    <!-- 屏幕 -->
    <link name="screen">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="5.0" />
            <inertia ixx="0.7" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/screen.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/screen.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 按钮 -->
    <link name="button">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="5.0" />
            <inertia ixx="0.7" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/button.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="red" />
        </visual>

        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/button.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 充电口 -->
    <link name="charge">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="5.0" />
            <inertia ixx="0.7" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/charge.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="green" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/charge.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 光通信口 -->
    <link name="sun">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <mass value="5.0" />
            <inertia ixx="0.7" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.5" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/sun.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="green" />
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-1.5707 0.0 -1.5707" />
            <geometry>
                <mesh filename="package://car_description/meshes/sun.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>


    <!-- 主舵轮 -->
    <link name="main_wheel_link">
        <inertial>
            <origin xyz="0 0.0 0.0" rpy="-1.5707 0.0 0" />
            <mass value="100.0" />
            <inertia ixx="50" ixy="0.0" ixz="0.0" iyy="42" iyz="0.0" izz="22" />
        </inertial>
        <visual>
            <origin xyz="0 0.0 0.0" rpy="-1.5707 0.0 0" />
            <geometry>
                <cylinder radius="0.105" length="0.075"/>
            </geometry>
            <material name="gray"/>
        </visual>
        <visual>
            <origin xyz="0 0.0 0.00" rpy="-1.5707 0.0 0" />
            <geometry>
                <cylinder radius="0.105" length="0.075"/>

            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0.0 0.0" rpy="-1.5707 0.0 0" />
            <geometry>
                <cylinder radius="0.105" length="0.075"/>

            </geometry>
        </collision>
    </link>
    <!-- 左前轮 -->
    <link name="left_wheel_link">
        <inertial>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <mass value="30.0" />
            <inertia ixx="10" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="12" />
        </inertial>
        <visual>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <geometry>
                <mesh filename="package://car_description/meshes/frontwheel.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="gray"/>
        </visual>
        <collision>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <geometry>
                <mesh filename="package://car_description/meshes/frontwheel.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>

    </link>
    <!-- 右前轮 -->
    <link name="right_wheel_link">
        <inertial>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <mass value="30.0" />
            <inertia ixx="10" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="12" />
        </inertial>
        <visual>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <geometry>
                <mesh filename="package://car_description/meshes/frontwheel.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
            <material name="gray"/>
        </visual>
        <collision>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <geometry>
                <mesh filename="package://car_description/meshes/frontwheel.stl" scale="0.0010 0.0010 0.0010" />
            </geometry>
        </collision>
    </link>
    <!-- 主轮转向轴 -->
    <link name="main_wheel_base_link">
        <inertial>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <mass value="100.0" />
            <inertia ixx="20" ixy="0.0" ixz="0.0" iyy="12" iyz="0.0" izz="12" />
        </inertial>
    </link>


    <!-- 导航2D雷达 -->
    <link name="main_2d_lidar_link">
        <inertial>
            <origin xyz="0 0.0 0.0" rpy="0 0.0 0" />
            <mass value="3.0" />
            <inertia ixx="0.04" ixy="0.0" ixz="0.0" iyy="0.04" iyz="0.0" izz="0.04" />
        </inertial>
    </link>

    <joint name="main_2d_lidar_foot_joint" type="fixed">
        <origin xyz="0.115 0.0 1.9" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="main_2d_lidar_link"/>
    </joint>

    <joint name="frame_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="fork_frame"/>
    </joint>
    <joint name="outface_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="out_face"/>
    </joint>
    <joint name="bar_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="bar"/>
    </joint>
    <joint name="charge_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="charge"/>
    </joint>
    <joint name="screen_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="screen"/>
    </joint>
    <joint name="button_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="button"/>
    </joint>
    <joint name="avoid_lidar_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="avoid_lidar"/>
    </joint>
    <joint name="sun_foot_joint" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="sun"/>
    </joint>


    <joint name="fork_joint" type="prismatic">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="fork"/>
        <axis xyz="0.0 0.0 1"/>
        <limit lower="0" upper="3.0" effort="20.0" velocity="2.0"/>
        <dynamics damping="1.0" friction="0.5"/>
    </joint>
    <joint name="main_wheel_base_joint" type="revolute">
        <origin xyz="0.21 0.0 0.04" rpy="0.0 0.0 0.0"/>
        <parent link="footbase"/>
        <child link="main_wheel_base_link"/>
        <axis xyz="0.0 0.0 1"/>
        <limit lower="-1.5707" upper="1.5707" effort="20.0" velocity="1.5707"/>
        <dynamics damping="1.0" friction="0.5"/>
    </joint>
    <joint name="main_wheel_joint" type="continuous">
        <origin xyz="0 0.0 0" rpy="0.0 0.0 0.0"/>
        <parent link="main_wheel_base_link"/>
        <child link="main_wheel_link"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="-100" upper="100.0" effort="20.0" velocity="2.0"/>
        <dynamics damping="1.0" friction="0.5"/>
    </joint>
    <joint name="left_wheel_joint" type="continuous">
        <origin xyz="-1.01 0.2525 -0.015" rpy="0.0 0.0 1.5707"/>
        <parent link="footbase"/>
        <child link="left_wheel_link"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-100" upper="100.0" effort="20.0" velocity="2.0"/>
        <dynamics damping="1.0" friction="0.5"/>
    </joint>
    <joint name="right_wheel_joint" type="continuous">
        <origin xyz="-1.01 -0.2525 -0.015" rpy="0.0 0.0 1.5707"/>
        <parent link="footbase"/>
        <child link="right_wheel_link"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-100" upper="100.0" effort="20.0" velocity="2.0"/>
        <dynamics damping="1.0" friction="0.5"/>
    </joint>
    <!-- ros2_control -->
    <ros2_control name="GazeboSimSystem" type="system">
        <hardware>
            <plugin>gz_ros2_control/GazeboSimSystem</plugin>
        </hardware>
        <joint name="fork_joint">
            <command_interface name="position">
                <param name="min">0</param>
                <param name="max">3</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
        <joint name="main_wheel_base_joint">
            <command_interface name="position">
                <param name="min">-1.5707</param>
                <param name="max">1.5707</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
        <joint name="main_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-30</param>
                <param name="max">30</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
    </ros2_control>
    <gazebo>
        <!-- Joint state publisher -->
        <plugin filename="libgz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
            <parameters>$(find car_description)/config/car.yaml</parameters>
        </plugin>
    </gazebo>

    <gazebo reference="main_2d_lidar_link">
        <sensor name='gpu_lidar' type='gpu_lidar'>
            <always_on>true</always_on>
            <visualize>false</visualize>
            <topic>lidar</topic>
            <update_rate>10</update_rate>
            <!-- 必须添加gz_frame_id 否则rviz2无法显示激光 -->
            <gz_frame_id>main_2d_lidar_link</gz_frame_id>
            <lidar>
                <scan>
                    <horizontal>
                        <samples>3600</samples>
                        <resolution>1</resolution>
                        <min_angle>-2.35619</min_angle>
                        <max_angle>2.35619</max_angle>
                    </horizontal>
                    <vertical>
                        <samples>1</samples>
                        <resolution>0.01</resolution>
                        <min_angle>0</min_angle>
                        <max_angle>0</max_angle>
                    </vertical>
                </scan>
                <range>
                    <min>0.08</min>
                    <max>10.0</max>
                    <resolution>0.01</resolution>
                </range>
                <noise>
                    <type>
                        gaussian
                    </type>
                    <mean>0</mean>
                    <stddev>
                        0.001
                    </stddev>
                </noise>
            </lidar>
        </sensor>
    </gazebo>


</robot>